src/theme-generator/vscode-theme-generator.ts

```diff
@@ -0,0 +1,435 @@
+import * as color from './color';
+import * as rules from './rules';
+
+import {IColorSet, IThemeGenerator} from './interfaces';
+import {IVscodeJsonThemeSetting} from './rules';
+
+export interface IVscodeJsonTheme {
+  name?: string;
+  semanticHighlighting?: boolean;
+  settings?: IVscodeJsonThemeSetting[];
+  tokenColors?: IVscodeJsonThemeSetting[];
+  colors?: {
+    [key: string]: string;
+  };
+}
+
+export class VscodeThemeGenerator implements IThemeGenerator {
+  generateTheme(name: string, colorSet: IColorSet): string {
+    // Fill in missing subsets to prevent NPEs
+    if (!colorSet.type) {
+      colorSet.type = 'dark';
+    }
+
+    if (!colorSet.syntax) {
+      colorSet.syntax = {};
+    }
+
+    if (!colorSet.terminal) {
+      colorSet.terminal = {};
+    }
+
+    if (!colorSet.ui) {
+      colorSet.ui = {};
+    }
+
+    const fallbackColorSet = color.generateFallbackColorSet(colorSet.base, colorSet.type, colorSet.semanticHighlighting);
+
+    const theme: IVscodeJsonTheme = {
+      name: name,
+      semanticHighlighting: colorSet.semanticHighlighting || false,
+      tokenColors: [],
+      colors: {}
+    };
+
+    const globalSettings: IVscodeJsonThemeSetting = {
+      name: 'Global settings',
+      settings: {},
+      scope: 'N/A'
+    };
+
+    theme.tokenColors.push(globalSettings);
+
+    rules.tokenRules.forEach(generator => {
+      const color = generator.color(colorSet) || generator.color(fallbackColorSet);
+      if (color) {
+        theme.tokenColors.push(generator.generate(color));
+      }
+    });
+
+    rules.globalRules.forEach(generator => {
+      const color = generator.color(colorSet) || generator.color(fallbackColorSet);
+      if (color) {
+        const generated = generator.generate(color);
+        // @ts-expect-error
```

Comment:
Would love someone's assistance here. Thanks!

Reply:
Typing on `Object.keys` always sucks, can cast that as an "as" (keyof typeof globalSettings.settings) maybe?